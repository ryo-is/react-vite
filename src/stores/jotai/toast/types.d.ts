export type Toast = {
  status: 'info' | 'error' | 'success' | 'hidden';
  message: string;
};
