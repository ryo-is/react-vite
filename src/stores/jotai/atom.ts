import { atom } from 'jotai';
import { atomWithStorage } from 'jotai/utils';

export const countAtom = atom({ count: 0 });
export const getCountValue = atom((get) => get(countAtom).count);
export const getDoubleCountValue = atom((get) => get(countAtom).count * 2);

export const countAtomWithStorageAtom = atomWithStorage(
  'countAtomWithStorage',
  0
);
