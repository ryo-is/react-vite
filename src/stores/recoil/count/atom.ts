import { atom } from 'recoil';

import type { InitialState } from './types';

const initialState: InitialState = {
  count: 0,
};

export const countAtom = atom<InitialState>({
  key: 'count_atom',
  default: initialState,
});

export const count2Atom = atom<InitialState>({
  key: 'count_2_atom',
  default: initialState,
});
