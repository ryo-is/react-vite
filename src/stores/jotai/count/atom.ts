import { atom } from 'jotai';
import { atomWithStorage } from 'jotai/utils';

export const countAtom = atom({ count: 0 });
export const count2Atom = atom({ count: 0 });
export const getCountValue = atom((get) => get(countAtom).count);
export const getCount2Value = atom((get) => get(count2Atom).count);
export const getDoubleCountValue = atom((get) => get(countAtom).count * 2);
export const getSumCountValue = atom(
  (get) => get(countAtom).count + get(count2Atom).count,
);

export const countAtomWithStorageAtom = atomWithStorage(
  'countAtomWithStorage',
  0,
);
