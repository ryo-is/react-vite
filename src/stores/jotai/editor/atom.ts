import { atom, useAtom } from 'jotai';

const DefaultValueIos = `<!DOCTYPE html>
<html lang="ja">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/@ionic/core@7/css/ionic.bundle.css"
    />
    <script src="https://cdn.jsdelivr.net/npm/@ionic/core@7/dist/ionic/ionic.js"></script>
    <title>ポップアップ</title>
  </head>
  <body>
    <style>
      ion-modal {
        --box-shadow: 0 10px 15px -3px rgb(0 0 0 / 0.1), 0 4px 6px -4px rgb(0 0 0 / 0.1);
      }
    
      ion-modal::part(backdrop) {
        background: rgba(209, 213, 219);
        opacity: 1;
      }
    
      ion-modal ion-toolbar {
        --background: rgb(14 116 144);
        --color: white;
      }
    </style>

    <ion-modal is-open="true" mode="ios" style="">
      <ion-header translucent>
        <ion-toolbar>
          <ion-title>ポップアップ</ion-title>
          <ion-buttons slot="end">
            <ion-button onclick="dismissModal()">閉じる</ion-button>
          </ion-buttons>
        </ion-toolbar>
      </ion-header>
      <ion-content >
        <ion-grid>
          <ion-row class="ion-padding">
            <ion-col size="4">
              <ion-img
                src="https://assets.stg.sprocket.bz/images/original/3f7c133e28c645a187d5d264ead5a54e/1-2.png"
              />
            </ion-col>
            <ion-col size="8">
              <ion-row>
                <ion-col size="12" class="ion-margin-bottom">
                  <ion-text>
                    <strong>商品検索をしてみませんか？</strong>
                  </ion-text>
                </ion-col>
                <ion-col size="12">
                  <ion-text
                    >欲しい商品がきっと見つかります</ion-text
                  >
                </ion-col>
              </ion-row>
            </ion-col>
            <ion-col size="6">
              <ion-button id="show_search" expand="block"
                >商品検索を使う</ion-button
              >
            </ion-col>
            <ion-col size="6">
              <ion-button id="dismiss" expand="block" color="light"
                >閉じる</ion-button
              >
            </ion-col>
          </ion-row>
        </ion-grid>
      </ion-content>
    </ion-modal>

    <script>
      var showSearchButton = document.getElementById("show_search");
      showSearchButton.addEventListener(
        "click",
        function () {
          sprocketSdk.trigger("button.click", "show_search");
        },
        false
      );

      var dismissButton = document.getElementById("dismiss");
      dismissButton.addEventListener(
        "click",
        function () {
          sprocketSdk.trigger("dismiss", "close");
        },
        false
      );
    </script>
  </body>
</html>`;

const DefaultValueAndoid = `<!DOCTYPE html>
<html lang="ja">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/@ionic/core@7/css/ionic.bundle.css"
    />
    <script src="https://cdn.jsdelivr.net/npm/@ionic/core@7/dist/ionic/ionic.js"></script>
    <title>ポップアップ</title>
  </head>
  <body>
    <style>
      ion-modal {
        --box-shadow: 0 10px 15px -3px rgb(0 0 0 / 0.1), 0 4px 6px -4px rgb(0 0 0 / 0.1);
      }
    
      ion-modal::part(backdrop) {
        background: rgba(209, 213, 219);
        opacity: 1;
      }
    
      ion-modal ion-toolbar {
        --background: rgb(14 116 144);
        --color: white;
      }
    </style>

    <ion-modal is-open="true" mode="md" style="">
      <ion-header translucent>
        <ion-toolbar>
          <ion-title>ポップアップ</ion-title>
          <ion-buttons slot="end">
            <ion-button onclick="dismissModal()">閉じる</ion-button>
          </ion-buttons>
        </ion-toolbar>
      </ion-header>
      <ion-content >
        <ion-grid>
          <ion-row class="ion-padding">
            <ion-col size="4">
              <ion-img
                src="https://assets.stg.sprocket.bz/images/original/3f7c133e28c645a187d5d264ead5a54e/1-2.png"
              />
            </ion-col>
            <ion-col size="8">
              <ion-row>
                <ion-col size="12" class="ion-margin-bottom">
                  <ion-text>
                    <strong>商品検索をしてみませんか？</strong>
                  </ion-text>
                </ion-col>
                <ion-col size="12">
                  <ion-text
                    >欲しい商品がきっと見つかります</ion-text
                  >
                </ion-col>
              </ion-row>
            </ion-col>
            <ion-col size="6">
              <ion-button id="show_search" expand="block"
                >商品検索を使う</ion-button
              >
            </ion-col>
            <ion-col size="6">
              <ion-button id="dismiss" expand="block" color="light"
                >閉じる</ion-button
              >
            </ion-col>
          </ion-row>
        </ion-grid>
      </ion-content>
    </ion-modal>

    <script>
      var showSearchButton = document.getElementById("show_search");
      showSearchButton.addEventListener(
        "click",
        function () {
          sprocketSdk.trigger("button.click", "show_search");
        },
        false
      );

      var dismissButton = document.getElementById("dismiss");
      dismissButton.addEventListener(
        "click",
        function () {
          sprocketSdk.trigger("dismiss", "close");
        },
        false
      );
    </script>
  </body>
</html>`;

export const editorContentValue = atom<string>(DefaultValueIos);
export const editorModalIsOpen = atom<boolean>(false);

export const useEditorContentValue = () => {
  const [content, setContent] = useAtom(editorContentValue);

  const setIosContent = () => {
    setContent(DefaultValueIos);
  };

  const setAndroidContent = () => {
    setContent(DefaultValueAndoid);
  };

  return {
    content,
    setContent,
    setIosContent,
    setAndroidContent,
  };
};

export const useEditorModalIsOpen = () => {
  const [isOpen, setIsOpen] = useAtom(editorModalIsOpen);

  return {
    isOpen,
    setIsOpen,
  };
};
