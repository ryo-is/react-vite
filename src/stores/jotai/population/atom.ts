import { atom } from 'jotai';
import { Population, Prefecture } from './types';

export const prefectureAtom = atom<Prefecture[]>([]);
export const selectedPrefectureAtom = atom<Prefecture[]>([]);
export const populationAtom = atom<Population[]>([]);

export const usePrefectures = atom(
  (get) => get(prefectureAtom),
  async (_, set) => {
    const res = await fetch(
      'https://opendata.resas-portal.go.jp/api/v1/prefectures',
      {
        headers: {
          'X-API-KEY': import.meta.env.VITE_API_KEY as string,
        },
      }
    );
    const { result, statusCode } = (await res.json()) as {
      result: Prefecture[];
      statusCode?: number;
      message: string;
    };
    if (!statusCode) {
      set(prefectureAtom, result);
    } else {
      set(prefectureAtom, []);
    }
  }
);

export const useSelectedPrefectures = atom(
  (get) => get(selectedPrefectureAtom),
  (_, set, { checked, pref }: { checked: boolean; pref: Prefecture }) => {
    if (checked) {
      set(selectedPrefectureAtom, (prev) => [...prev, pref]);
    } else {
      set(selectedPrefectureAtom, (prev) => {
        const prevPref = [...prev];
        const idx = prevPref.findIndex((p) => p.prefCode === pref.prefCode);
        if (idx >= 0) {
          prevPref.splice(idx, 1);
        }
        return prevPref;
      });
    }
  }
);

export const usePopulations = atom(
  (get) => get(populationAtom),
  async (get, set) => {
    const targetPrefs = get(selectedPrefectureAtom);
    if (targetPrefs.length <= 0) {
      set(populationAtom, () => []);
    }
    const promises = targetPrefs.map((pref) =>
      fetch(
        `https://opendata.resas-portal.go.jp/api/v1/population/composition/perYear?cityCode=-&prefCode=${pref.prefCode}`,
        {
          headers: {
            'X-API-KEY': import.meta.env.VITE_API_KEY as string,
          },
        }
      )
    );
    const results = await Promise.all(promises);
    const toJsonPromises = results.map((r) => r.json());
    const populationJsons = await Promise.all<{
      result: {
        data: { data: { year: number; value: number }[] }[];
      };
      statusCode: number;
      message: string;
    }>(toJsonPromises);
    const errorResult = populationJsons.find(
      (json) => json.statusCode !== undefined
    );
    if (!errorResult) {
      const populations: Population[] = populationJsons.map((p, i) => ({
        ...p.result.data[0],
        label: targetPrefs[i].prefName,
      }));
      set(populationAtom, populations);
    } else {
      set(populationAtom, () => []);
    }
  }
);

export const usePopulationSeries = atom((get) => {
  const populations = get(populationAtom);
  return populations.map(({ label, data }) => ({
    label,
    data: data.map((d) => ({
      ...d,
    })),
  }));
});
