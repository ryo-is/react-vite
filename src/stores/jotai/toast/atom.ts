import { atom } from 'jotai';

import { Toast } from './types';

const initialToastAtom: Toast = { status: 'hidden', message: '' };
export const toastAtom = atom<Toast>(initialToastAtom);

export const useToast = atom(
  (get) => get(toastAtom),
  (_, set, toast: Toast) => {
    set(toastAtom, toast);
    setTimeout(() => set(toastAtom, initialToastAtom), 3000);
  },
);
