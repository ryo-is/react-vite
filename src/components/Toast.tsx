import { ExclamationCircleIcon } from '@heroicons/react/20/solid';
import { useAtom } from 'jotai';
import { useToast } from '../stores/jotai/toast/atom';

const alertLevel: Record<'info' | 'error' | 'success' | 'hidden', string> = {
  info: 'alert-info',
  error: 'alert-error',
  success: 'alert-success',
  hidden: '',
};

export const Toast = () => {
  const [toast] = useAtom(useToast);

  if (toast.status === 'hidden') {
    return <div className="hidden" />;
  }

  return (
    <div className="toast">
      <div className={`alert ${alertLevel[toast.status]}`}>
        <div className="flex gap-x-1 px-2">
          {toast.status === 'error' && (
            <ExclamationCircleIcon className="mr-1 h-6 w-6" />
          )}
          <span>{toast.message}</span>
        </div>
      </div>
    </div>
  );
};
