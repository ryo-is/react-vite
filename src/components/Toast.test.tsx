import { render, screen } from '@testing-library/react';
import { useHydrateAtoms } from 'jotai/utils';
import { Provider } from 'jotai';
import { ReactNode } from 'react';
import { Toast } from './Toast';
import { Toast as ToastType } from '../stores/jotai/toast/types';
import { toastAtom } from '../stores/jotai/toast/atom';

type ProviderProps = {
  initialValues: [typeof toastAtom, ToastType][];
  children: ReactNode;
};

const HydrateAtoms = ({ initialValues, children }: ProviderProps) => {
  useHydrateAtoms(initialValues);
  return children;
};

const TestProvider = ({ initialValues, children }: ProviderProps) => (
  <Provider>
    <HydrateAtoms initialValues={initialValues}>{children}</HydrateAtoms>
  </Provider>
);

const ToastProvider = () => (
  <TestProvider
    initialValues={[[toastAtom, { status: 'info', message: 'test' }]]}
  >
    <Toast />
  </TestProvider>
);

describe('Toast', () => {
  it('should match to the snapshot', () => {
    const { asFragment } = render(<Toast />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display message is test', () => {
    render(<ToastProvider />);

    const message = screen.getByTestId('toast-message');
    expect(message.textContent).toEqual('test');
  });
});
