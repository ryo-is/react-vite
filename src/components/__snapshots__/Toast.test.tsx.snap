// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toast > should match to the snapshot 1`] = `
<DocumentFragment>
  <div
    class="hidden"
  />
</DocumentFragment>
`;
